File: cdn/cloud_api/pkg/service/auth/resource.go
package auth

import cloudauth "a.yandex-team.ru/cloud/iam/accessservice/client/iam-access-service-client-go/v1"

type AuthorizeResource interface {
	IsAuthorizeResource()
}

type CloudAuthResource interface {
	GetResources() []cloudauth.Resource
}

type Folder struct {
	FolderID string
}

func (f *Folder) IsAuthorizeResource() {}

func (f *Folder) GetResources() []cloudauth.Resource {
	return []cloudauth.Resource{cloudauth.ResourceFolder(f.FolderID)}
}

// TODO: not used
type Cloud struct {
	CloudID string
}

func (c *Cloud) IsAuthorizeResource() {}

func (c *Cloud) GetResources() []cloudauth.Resource {
	return []cloudauth.Resource{cloudauth.ResourceCloud(c.CloudID)}
}

type Entity struct {
	FolderID   string
	EntityID   string
	EntityType EntityType
}

func (e *Entity) IsAuthorizeResource() {}

func (e *Entity) GetResources() []cloudauth.Resource {
	return []cloudauth.Resource{
		{
			ID:   e.EntityID,
			Type: e.EntityType.String(),
		},
		cloudauth.ResourceFolder(e.FolderID),
	}
}

type EntityType string

func (t EntityType) String() string {
	return string(t)
}

const (
	CDNResourceType EntityType = "cdn.resource"
)

File: cloud/iam/accessservice/client/iam-access-service-client-cpp/1.0.0/src/resource.cpp
#include "cloud-auth/resource.h"

#include <iostream>

namespace yandex::cloud::auth {

    static const auto RESOURCE_TYPE_CLOUD = "resource-manager.cloud";
    static const auto RESOURCE_TYPE_FOLDER = "resource-manager.folder";

    std::ostream& operator<<(std::ostream& out, const Resource& resource) {
        return out << "Resource {id='" << resource.Id() << "', type='" << resource.Type() << "'}";
    }

    Resource Resource::Cloud(std::string id) {
        return {std::move(id), RESOURCE_TYPE_CLOUD};
    }

    Resource Resource::Folder(std::string id) {
        return {std::move(id), RESOURCE_TYPE_FOLDER};
    }

}
